File: android/src/main/java/expo/modules/swipeable/SwipeableModule.kt
package expo.modules.swipeable

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class SwipeableModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("Swipeable")

        // Module-level functions (by recyclingKey)
        Function("openByKey") { key: String ->
            SwipeableView.openByKey(key)
        }

        Function("closeByKey") { key: String, animated: Boolean? ->
            SwipeableView.closeByKey(key, animated ?: true)
        }

        Function("cancelByKey") { key: String ->
            SwipeableView.cancelByKey(key)
        }

        Function("closeAll") { animated: Boolean? ->
            SwipeableView.closeAll(animated ?: true)
        }

        Function("isOpenByKey") { key: String ->
            SwipeableView.isOpenByKey(key)
        }

        View(SwipeableView::class) {
            Prop("actionsWidth") { view: SwipeableView, width: Float ->
                view.actionsWidth = width
            }
            Prop("actionsPosition") { view: SwipeableView, position: String ->
                view.actionsPosition = position
            }
            Prop("friction") { view: SwipeableView, friction: Float ->
                view.friction = friction.coerceIn(0f, 1f)
            }
            Prop("threshold") { view: SwipeableView, threshold: Float ->
                view.threshold = threshold.coerceIn(0f, 1f)
            }
            Prop("gestureEnabled") { view: SwipeableView, enabled: Boolean ->
                view.gestureEnabled = enabled
            }
            Prop("dragOffsetFromEdge") { view: SwipeableView, offset: Float ->
                view.dragOffsetFromEdge = offset
            }
            Prop("recyclingKey") { view: SwipeableView, key: String? ->
                view.recyclingKey = key
            }
            Prop("autoClose") { view: SwipeableView, autoClose: Boolean ->
                view.autoClose = autoClose
            }
            Prop("autoCloseTimeout") { view: SwipeableView, timeout: Float ->
                view.autoCloseTimeout = timeout
            }

            Events("onSwipeProgress", "onSwipeStart", "onSwipeStateChange", "onSwipeEnd")

            AsyncFunction("close") { view: SwipeableView, animated: Boolean? ->
                view.post { view.close(animated ?: true) }
            }
            AsyncFunction("open") { view: SwipeableView ->
                view.post { view.open() }
            }
            AsyncFunction("cancel") { view: SwipeableView ->
                view.post { view.cancelGesture() }
            }

            AsyncFunction("handleGestureStart") { view: SwipeableView ->
                view.handleGestureStart()
            }
            AsyncFunction("handleGestureUpdate") { view: SwipeableView, translationX: Float, velocityX: Float ->
                view.handleGestureUpdate(translationX, velocityX)
            }
            AsyncFunction("handleGestureEnd") { view: SwipeableView, translationX: Float, velocityX: Float ->
                view.handleGestureEnd(translationX, velocityX)
            }
        }
    }
}

File: ios/SwipeableModule.swift
import ExpoModulesCore

public class SwipeableModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Swipeable")

        // Module-level functions (by recyclingKey)
        Function("openByKey") { (key: String) in
            SwipeableView.openByKey(key)
        }

        Function("closeByKey") { (key: String, animated: Bool?) in
            SwipeableView.closeByKey(key, animated: animated ?? true)
        }

        Function("cancelByKey") { (key: String) in
            SwipeableView.cancelByKey(key)
        }

        Function("closeAll") { (animated: Bool?) in
            SwipeableView.closeAll(animated: animated ?? true)
        }

        Function("isOpenByKey") { (key: String) -> Bool in
            SwipeableView.getOpenState(for: key)
        }

        View(SwipeableView.self) {
            // Props
            Prop("actionsWidth") { (view: SwipeableView, width: CGFloat) in
                view.actionsWidth = width
            }

            Prop("actionsPosition") { (view: SwipeableView, position: String) in
                view.actionsPosition = position
            }

            Prop("friction") { (view: SwipeableView, friction: CGFloat) in
                view.friction = friction
            }

            Prop("threshold") { (view: SwipeableView, threshold: CGFloat) in
                view.threshold = threshold
            }

            Prop("gestureEnabled") { (view: SwipeableView, enabled: Bool) in
                view.gestureEnabled = enabled
            }

            Prop("dragOffsetFromEdge") { (view: SwipeableView, offset: CGFloat) in
                view.dragOffsetFromEdge = offset
            }

            Prop("recyclingKey") { (view: SwipeableView, key: String?) in
                view.recyclingKey = key
            }

            Prop("autoClose") { (view: SwipeableView, autoClose: Bool) in
                view.autoClose = autoClose
            }

            Prop("autoCloseTimeout") { (view: SwipeableView, timeout: CGFloat) in
                view.autoCloseTimeout = timeout
            }

            // Events
            Events("onSwipeProgress", "onSwipeStart", "onSwipeStateChange", "onSwipeEnd")

            // Imperative methods
            AsyncFunction("close") { (view: SwipeableView, animated: Bool?) in
                DispatchQueue.main.async {
                    view.close(animated: animated ?? true)
                }
            }

            AsyncFunction("open") { (view: SwipeableView) in
                DispatchQueue.main.async {
                    view.open()
                }
            }

            AsyncFunction("cancel") { (view: SwipeableView) in
                DispatchQueue.main.async {
                    view.cancelGesture()
                }
            }

            // Gesture handling methods (called from JS gesture handler)
            AsyncFunction("handleGestureStart") { (view: SwipeableView) in
                view.handleGestureStart()
            }

            AsyncFunction("handleGestureUpdate") { (view: SwipeableView, translationX: CGFloat, velocityX: CGFloat) in
                view.handleGestureUpdate(translationX: translationX, velocityX: velocityX)
            }

            AsyncFunction("handleGestureEnd") { (view: SwipeableView, translationX: CGFloat, velocityX: CGFloat) in
                view.handleGestureEnd(translationX: translationX, velocityX: velocityX)
            }
        }
    }
}
